Classes/Plugin/OaiPmh.php

```diff
@@ -449,35 +456,43 @@ protected function resume()
      */
     protected function verbGetRecord()
     {
-        if (count($this->piVars) != 3 || empty($this->piVars['metadataPrefix']) || empty($this->piVars['identifier'])) {
+        if (count($this->piVars) !== 3 || empty($this->piVars['metadataPrefix']) || empty($this->piVars['identifier'])) {
             return $this->error('badArgument');
         }
-        if (!in_array($this->piVars['metadataPrefix'], array_keys($this->formats))) {
+        if (!array_key_exists($this->piVars['metadataPrefix'], $this->formats)) {
             return $this->error('cannotDisseminateFormat');
         }
         $where = '';
         if (!$this->conf['show_userdefined']) {
             $where .= ' AND tx_dlf_collections.fe_cruser_id=0';
         }
-        $record = $GLOBALS['TYPO3_DB']->exec_SELECT_mm_query(
-            'tx_dlf_documents.*,GROUP_CONCAT(DISTINCT tx_dlf_collections.oai_name ORDER BY tx_dlf_collections.oai_name SEPARATOR " ") AS collections',
-            'tx_dlf_documents',
-            'tx_dlf_relations',
-            'tx_dlf_collections',
-            'AND tx_dlf_documents.record_id=' . $GLOBALS['TYPO3_DB']->fullQuoteStr($this->piVars['identifier'], 'tx_dlf_documents')
-                . ' AND tx_dlf_documents.pid=' . intval($this->conf['pages'])
-                . ' AND tx_dlf_collections.pid=' . intval($this->conf['pages'])
-                . ' AND tx_dlf_relations.ident=' . $GLOBALS['TYPO3_DB']->fullQuoteStr('docs_colls', 'tx_dlf_relations')
-                . $where
-                . Helper::whereClause('tx_dlf_collections'),
-            '',
-            '',
-            '1'
-        );
-        if (!$GLOBALS['TYPO3_DB']->sql_num_rows($record)) {
+
+        $connection = GeneralUtility::makeInstance(ConnectionPool::class)->getConnectionForTable('tx_dlf_documents');
+
+        $sql = 'SELECT `tx_dlf_documents`.*, GROUP_CONCAT(DISTINCT `tx_dlf_collections`.oai_name ORDER BY tx_dlf_collections.oai_name SEPARATOR " ") AS `collections` ' .
+            'FROM `tx_dlf_documents` ' .
+            'INNER JOIN `tx_dlf_relations` ON `tx_dlf_relations`.`uid_local` = `tx_dlf_documents`.`uid` ' .
+            'INNER JOIN `tx_dlf_collections` ON `tx_dlf_collections`.`uid` = `tx_dlf_relations`.`uid_foreign` ' .
+            'WHERE (((`tx_dlf_documents`.`deleted` = 0) AND (`tx_dlf_collections`.`deleted` = 0)) AND ((`tx_dlf_documents`.`hidden` = 0) AND (`tx_dlf_collections`.`hidden` = 0)) AND (`tx_dlf_documents`.`starttime` <= ' . $GLOBALS['EXEC_TIME'] . ') AND ((`tx_dlf_documents`.`endtime` = 0) OR (`tx_dlf_documents`.`endtime` > ' . $GLOBALS['EXEC_TIME'] . '))) ' .
+            'AND tx_dlf_documents.record_id = ? ' .
+            'AND tx_dlf_documents.pid = ? ' .
+            'AND tx_dlf_collections.pid = ? ' .
+            'AND tx_dlf_relations.ident="docs_colls"' .
+            $where .
+            Helper::whereClause('tx_dlf_collections', [$this->piVars['identifier']]);
```

Comment:
The second argument of `Helper::whereClause()` has to be a boolean to define if hidden records should be shown.

Also, in order to use the QueryBuilder, using `Helper::whereExpression()` is prefered over using `Helper::whereClause()` (which uses the old DB API functions).

Reply:
I disagree on using `Helper::whereClause()` at all. It adds similar restriction constraints as QueryBuilder did already. This results is an unneeded complex SQL-query IMHO.

Example of a non working (other problem in line 795) Query with and without  `Helper::whereClause()`:

### with  `Helper::whereClause()`
<pre>
SELECT `tx_dlf_collections`.`oai_name` AS `oai_name`, `tx_dlf_collections`.`label` AS `label` 
FROM `tx_dlf_collections` 
WHERE (`tx_dlf_collections`.`sys_language_uid` IN (-1, 0)) 
AND (`tx_dlf_collections`.`pid` = 176) 
AND (`tx_dlf_collections`.`oai_name` <> ) 
AND (`tx_dlf_collections`.`fe_cruser_id` = 0) 
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0) 
  AND (
    ( 
      (`tx_dlf_collections`.`fe_group` = '') 
      OR (`tx_dlf_collections`.`fe_group` IS NULL) 
      OR (`tx_dlf_collections`.`fe_group` = '0') 
      OR (FIND_IN_SET('0', `tx_dlf_collections`.`fe_group`))
      OR (FIND_IN_SET('-1', `tx_dlf_collections`.`fe_group`)) 
     )
   )
)
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0)
 )
ORDER BY `tx_dlf_collections`.`oai_name` ASC
</pre>

### without  `Helper::whereClause()`
<pre>
SELECT `tx_dlf_collections`.`oai_name` AS `oai_name`, `tx_dlf_collections`.`label` AS `label` 
FROM `tx_dlf_collections` 
WHERE (`tx_dlf_collections`.`sys_language_uid` IN (-1, 0)) 
AND (`tx_dlf_collections`.`pid` = 176) 
AND (`tx_dlf_collections`.`oai_name` <> ) 
AND (`tx_dlf_collections`.`fe_cruser_id` = 0) 
AND (
  (`tx_dlf_collections`.`deleted` = 0) 
  AND (`tx_dlf_collections`.`hidden` = 0)
) 
ORDER BY `tx_dlf_collections`.`oai_name` ASC
</pre>